stellar-contract-env-common/src/val_wrapper.rs

```diff
@@ -0,0 +1,151 @@
+#[macro_export]
+macro_rules! decl_tagged_val_wrapper {
+    ($tagname:ident) => {
+        // Declare the wrapper type.
+        #[derive(Copy, Clone)]
+        pub struct $tagname(RawVal);
+
+        // AsRef / AsMut to RawVal.
+        impl AsRef<RawVal> for $tagname {
+            fn as_ref(&self) -> &RawVal {
+                &self.0
+            }
+        }
+        impl AsMut<RawVal> for $tagname {
+            fn as_mut(&mut self) -> &mut RawVal {
+                &mut self.0
+            }
+        }
+        // Basic conversion support: wrapper to raw, and try-into helper.
+        impl From<$tagname> for RawVal {
+            fn from(b: $tagname) -> Self {
+                b.0
+            }
+        }
+        impl crate::RawValConvertible for $tagname {
+            #[inline(always)]
+            fn is_val_type(v: RawVal) -> bool {
+                v.has_tag(Tag::$tagname)
+            }
+            #[inline(always)]
+            unsafe fn unchecked_from_val(v: RawVal) -> Self {
+                $tagname(v)
+            }
+        }
+
+        // EnvVal ref/mut/from support
+        impl<E: Env> AsRef<$tagname> for EnvVal<E, $tagname> {
+            fn as_ref(&self) -> &$tagname {
+                &self.val
+            }
+        }
+        impl<E: Env> AsMut<$tagname> for EnvVal<E, $tagname> {
+            fn as_mut(&mut self) -> &mut $tagname {
+                &mut self.val
+            }
+        }
+        impl<E: Env> From<EnvVal<E, $tagname>> for EnvVal<E, RawVal> {
+            fn from(ev: EnvVal<E, $tagname>) -> Self {
+                EnvVal {
+                    env: ev.env,
+                    val: ev.val.to_raw(),
+                }
+            }
+        }
+        impl<E: Env> From<EnvVal<E, $tagname>> for RawVal {
+            fn from(ev: EnvVal<E, $tagname>) -> Self {
+                ev.val.0
+            }
+        }
```

Comment:
I think we need some more conversions:
- `impl From<$tagname> for RawVal`
- `impl<E: Env> TryIntoVal<E, RawVal> for $tagname`

Reply:
Oh, I see the first one further up, so no worries.

Reply:
Which also means the second one should be blanket impl.

Reply:
I was about to add these but ..

`From<$tagname> for RawVal` is already implemented up on line 20.

and I _think_ `TryIntoVal<E, RawVal>` comes for free from the fact that raw_val.rs line 121 has a `impl<E: Env> TryFrom<EnvVal<E, RawVal>> for $T` and then you blanket-impl `TryIntoVal` on env_val.rs line 94.

IOW I think they're already covered. If you find others you need, by all means.